core/runtime/core_builtin.odin

```diff
@@ -536,6 +557,54 @@ resize_dynamic_array :: proc(array: ^$T/[dynamic]$E, length: int, loc := #caller
 	return true
 }
 
+/*
+	Shrinks the capacity of a dynamic array down to the current length, or the given capacity.
+
+	If `new_cap` is -1, then `len(array)` is used.
+
+	Does nothing if `new_cap > cap(array)`.
+
+	If `new_cap > len(array)`, then `len(array)` will be left unchanged.
+*/
+shrink_dynamic_array :: proc(array: ^$T/[dynamic]$E, new_cap := -1, loc := #caller_location) -> bool {
+	if array == nil {
+		return false
+	}
+	a := (^Raw_Dynamic_Array)(array)
+
+	new_cap := new_cap if new_cap != -1 else a.len
+
+	if new_cap > a.cap {
+		return true
```

Comment:
Should this be `true`?

Reply:
I figured that it should report that it was successful if there was no shrinking required, since you'd generally only report an error if it was unable to perform its job.

Reply:
I'm not sure if it's useful to distinquish the third one or not. 🤔 
If we want to do that though, maybe we could do `-> (did_shrink: bool, err: Allocator_Error)` ?

Reply:
Could always return `(did_shrink: bool, err: Allocator_Error)` instead I suppose. That actually seems easy enough to do.